compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirCallableMemberDeclaration
+import org.jetbrains.kotlin.fir.declarations.FirClass
+import org.jetbrains.kotlin.fir.declarations.modality
+
+object FirOpenMemberChecker : FirClassChecker() {
+    override fun check(declaration: FirClass<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration.canHaveOpenMembers()) return
+        for (memberDeclaration in declaration.declarations) {
+            if (memberDeclaration !is FirCallableMemberDeclaration<*>) continue
+            val source = memberDeclaration.source ?: continue
+            if (memberDeclaration.modality == Modality.OPEN) {
```

Comment:
Sorry for not catching this earlier. If a certain declaration has many, conflicting modalities, it's not guaranteed that checking this is enough to pick up `open` members. So you need to something like this:
```kt
// If multiple (potentially conflicting) modality modifiers are specified, not all modifiers are recorded at `status`.
// So, our source of truth should be the full modifier list retrieved from the source.
val modifierList = with(FirModifierList) { source.getModifierList() }
val hasOpenModifier = modifierList?.modifiers?.any { it.token == KtTokens.OPEN_KEYWORD } == true
val isOpen = memberDelcaration.isOpen || hasOpenModifier
```

Btw, you can use `FirMemberDeclaration.isOpen` which does that modality check.

Reply:
OK, let me filter on the type of fake source element kind.